src/libbson/src/bson/bson-context.c

```diff
@@ -264,79 +142,236 @@ _bson_context_get_hostname (char *out)
-      /* We mask off the last nibble so that the last digit of the OID will
-       * start at zero. Just to be nice. */
-      context->seq32 = _get_rand (&seed) & 0x007FFFF0;
+   switch (left) {
+   case 7:
+      b |= ((uint64_t) ni[6]) << 48;
+      /* FALLTHRU */
+   case 6:
+      b |= ((uint64_t) ni[5]) << 40;
+      /* FALLTHRU */
+   case 5:
+      b |= ((uint64_t) ni[4]) << 32;
+      /* FALLTHRU */
+   case 4:
+      b |= ((uint64_t) ni[3]) << 24;
+      /* FALLTHRU */
+   case 3:
+      b |= ((uint64_t) ni[2]) << 16;
+      /* FALLTHRU */
+   case 2:
+      b |= ((uint64_t) ni[1]) << 8;
+      /* FALLTHRU */
+   case 1:
+      b |= ((uint64_t) ni[0]);
+      break;
+   default:
+      BSON_UNREACHABLE ("Invalid remainder during SipHash");
+   case 0:
+      break;
    }
 
-   rand_bytes = _get_rand (&seed);
-   rand_bytes <<= 32;
-   rand_bytes |= _get_rand (&seed);
+   v3 ^= b;
+
+   for (i = 0; i < C_ROUNDS; ++i)
+      _sip_round (&v0, &v1, &v2, &v3);
+
+   v0 ^= b;
+
+   v2 ^= 0xee;
+
+   for (i = 0; i < D_ROUNDS; ++i)
+      _sip_round (&v0, &v1, &v2, &v3);
+
+   b = v0 ^ v1 ^ v2 ^ v3;
+   U64TO8_LE (digest_buf, b);
+
+   v1 ^= 0xdd;
+
+   for (i = 0; i < D_ROUNDS; ++i)
+      _sip_round (&v0, &v1, &v2, &v3);
+
+   b = v0 ^ v1 ^ v2 ^ v3;
+   U64TO8_LE (digest_buf + 8, b);
 
-   /* Copy five random bytes, endianness does not matter. */
-   memcpy (&context->rand, (char *) &rand_bytes, sizeof (context->rand));
+   memcpy (digest, digest_buf, sizeof digest_buf);
+}
+
+/*
+ * The seed consists of the following hashed together:
+ * - current time (with microsecond resolution)
+ * - current pid
+ * - current hostname
+ * - The init-call counter
+ */
+struct _init_rand_params {
+   struct timeval time;
+   uint64_t pid;
+   char hostname[HOST_NAME_MAX];
+   int64_t rand_call_counter;
+};
+
+static void
+_bson_context_init_random (bson_context_t *context, bool init_seq)
+{
+   /* The message digest of the random params */
+   uint64_t digest[2] = {0};
+   uint64_t key[2] = {0};
+   /* The randomness parameters */
+   struct _init_rand_params rand_params;
+
+   /* Init each part of the randomness source: */
+   memset (&rand_params, 0, sizeof rand_params);
+   bson_gettimeofday (&rand_params.time);
+   rand_params.pid = _bson_getpid ();
+   context->gethostname (rand_params.hostname);
+   rand_params.rand_call_counter =
+      bson_atomic_int64_fetch_add (&gRandCounter, 1, bson_memory_order_seq_cst);
+
+   memcpy (key, &rand_params, sizeof key);
+   key[1] = ~key[0];
+
+   /* Hash the param struct */
+   _siphash (&rand_params, sizeof rand_params, key, digest);
+
+   /** Initialize the rand and sequence counters with our random digest */
+   memcpy (context->randomness, digest, sizeof context->randomness);
+   if (init_seq) {
+      memcpy (&context->seq32, digest + 1, sizeof context->seq32);
+      memcpy (&context->seq64, digest + 1, sizeof context->seq64);
+      /* Chop off some initial bits for nicer counter behavior */
```

Comment:
What is "nicer counter behavior"? Is it "so that the last digit of the OID will start at zero" (per older comments)? If so, suggest this description is preserved.